server/gfmd/auth_uid.c

```diff
@@ -313,6 +344,46 @@ auth_uid_to_global_username_sharedsecret(void *closure,
 	return (GFARM_ERR_NO_ERROR);
 }
 
+#ifdef HAVE_CYRUS_SASL
+
+static gfarm_error_t
+auth_uid_to_global_username_sasl(void *closure,
+	const char *auth_user_id, enum gfarm_auth_id_role *auth_user_id_rolep,
+	char **global_usernamep)
+{
+	enum gfarm_auth_id_role auth_user_id_role = *auth_user_id_rolep;
+	char *global_username;
+	struct user *u;
+	const char diag[] = "auth_uid_to_global_username_sasl";
+
+	if (auth_user_id_role != GFARM_AUTH_ID_ROLE_USER)
+		return (GFARM_ERR_AUTHENTICATION);
+
+	giant_lock();
+	u = user_tenant_lookup(auth_user_id);
```

Comment:
This function should convert a sasl_user name in auth_user_id to a gfarm global username in *global_user_namep. But this implementation converts a gfarm global username in auth_user_id to a sasl_user name which is returned by user_auth_id() instead. In other words, the conversion is performed in the opposite direction than it should be.

Note that server/gfmd/auth_uid.c should do just same thing with lib/libgfarm/gfarm/auth_server_uid.c, the difference between them is:
* lib/libgfarm/gfarm/auth_server_uid.c uses a RPC to gfmd for the auth_user_id -> gfarm global user name conversion.
* server/gfmd/auth_uid.c uses a function call for the auth_user_id -> gfarm global user name conversion.